Source/Artifacts/Artifacts.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Subjects;
+using Microsoft.Extensions.Logging;
+
+namespace Dolittle.SDK.Artifacts
+{
+    /// <summary>
+    /// Represents an implementation of <see cref="IArtifacts" />.
+    /// </summary>
+    public class Artifacts : IArtifacts, IDisposable
+    {
+        readonly ReplaySubject<ArtifactAssociation> _registered;
+        readonly BehaviorSubject<IDictionary<Type, Artifact>> _associations;
+        readonly ILogger<Artifacts> _logger;
+        bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Artifacts"/> class.
+        /// </summary>
+        /// <param name="associations">The <see cref="IDictionary{TKey, TValue}"> artifact associations </see>.</param>
+        /// <param name="logger">The <see cref="ILogger" />.</param>
+        public Artifacts(IDictionary<Type, Artifact> associations, ILogger<Artifacts> logger)
+        {
+            _registered = new ReplaySubject<ArtifactAssociation>();
+            _associations = new BehaviorSubject<IDictionary<Type, Artifact>>(new Dictionary<Type, Artifact>());
+            _logger = logger;
+
+            foreach ((var type, var artifact) in associations)
+            {
+                _registered.OnNext(new ArtifactAssociation(type, artifact));
+            }
+
+            _registered.Subscribe(AddAssociation);
+        }
+
+        /// <inheritdoc />
+        public void Associate(Type type, Artifact artifact)
+            => _registered.OnNext(new ArtifactAssociation(type, artifact));
+
+        /// <inheritdoc />
+        public void Associate(Type type, ArtifactId artifactId)
+            => Associate(type, new Artifact(artifactId, Generation.First));
+
+        /// <inheritdoc />
+        public void Associate(Type type, ArtifactId artifactId, Generation generation)
+            => Associate(type, new Artifact(artifactId, generation));
+
+        /// <inheritdoc />
+        public Artifact GetFor<T>() => GetFor(typeof(T));
+
+        /// <inheritdoc />
+        public Artifact GetFor(Type type)
+        {
+            if (!_associations.Value.TryGetValue(type, out var artifact)) throw new UnknownArtifact(type);
+            return artifact;
+        }
+
+        /// <inheritdoc />
+        public Type GetTypeFor(Artifact artifact)
+        {
+            var type = _associations.Value.FirstOrDefault(_ => _.Value == artifact).Key;
+            if (type == default) throw new UnknownType(artifact);
+            return type;
+        }
+
+        /// <inheritdoc />
+        public Type GetTypeFor(ArtifactId artifactId) => GetTypeFor(new Artifact(artifactId));
```

Comment:
I think this method should go - as it doesn't really make sense to get a type for just an _id_. It doesn't have a type, only when you specify that it is the first generation - which feels very hidden here.

Reply:
Agree. I just mirrored the JS SDK